### ISSUE TITLE
Global Discovery User (discuss)

### ISSUE THREAD

**Issue by mekefly** [nostr-protocol/nips#790]
### Motivation

As the number of relays grows, for example to millions, we cannot have users send their 10002 to all relays. This would waste a lot of storage, waste a lot of compute, and also cannot guarantee that the user will be found.

### Prerequisites

* Global active addressing relay list 
	* to be maintained by relays
	* fallback 1: kind10002
	* fallback 2: kind3
* 10002 (nip-65)

### Basic Idea

Compare the hash value of the global active relay and the user's public key using bitwise XOR. The smaller the binary number, the higher the priority.

### Specific Steps

1. Get the global active relay list.
2. Convert the user's public key to a binary string.
3. Convert the hash value of the global active relay to a binary string.
4. Perform bitwise XOR on the binary string of the user's public key and the binary string of the hash value of the global active addressing relay.
5. Sort the results of the bitwise XOR operation to get the priority list.
6. Send the user's kind10002 to the top n relays.

### Complexity

* Algorithm network registration lookup complexity O1
* Algorithm network storage complexity O1

### For example
|Relay|User pubkey|Priority|
|---|---|---|
|1100|1100|0000|
|1101|1100|0001|
|1110|1100|0010|
|1000|1100|0100|
|1001|1100|0101|
|1010|1100|0110|
|0100|1100|1000|
|0000|1100|1100|
|0011|1100|1111|

**Comment by arthurfranca** [comment]
A bit offtopic but.. never understood why 10002 is using `r` tags instead of `relay` tags. Why index all these values bloating relays' DBs?

**Comment by mikedilger** [comment]
What is a Global active relay? Is it the URL? I don't like any global lists. If it's the URL there doesn't need to be a global list. Anyone can take all the relays they know about and to this operation, and start asking from the top down.

> A bit offtopic but.. never understood why 10002 is using r tags instead of relay tags. Why index all these values bloating relays' DBs?

I didn't notice that I did that. I think it was a mistake.

**Comment by mekefly** [comment]
> What is a Global active relay? Is it the URL? I don't like any global lists. If it's the URL there doesn't need to be a global list. Anyone can take all the relays they know about and do this operation, and start asking from the top down.
> 
> > A bit offtopic but.. never understood why 10002 is using r tags instead of relay tags. Why index all these values bloating relays' DBs?
> 
> I didn't notice that I did that. I think it was a mistake.

Indeed, this is possible. So my second consideration is to use kind10002 to get the list. After we get the relay list, we can upload it to more relays. However, the possibility of finding the user is still not 100%, which is also a drawback of this method. If you have a better idea, we can discuss it.

**Comment by arthurfranca** [comment]
> I didn't notice that I did that. I think it was a mistake.

@mikedilger Maybe it would be a good idea to slowly phase out `r` tag in favor of `relay` one editing the NIP text, what do you think? That said, thx for NIP-65 it is one of the most important NIPs.

**Comment by mikedilger** [comment]
It is important that if we change how kind 10002 tags are parsed, that we don't break existing clients that both create them the old way and parse them the old way.  So if we added a new tag "relay" and maybe even upgraded the markers to "inbox" and "outbox" to be clearer (read and write are too directional), .... we would still need to keep the 'r' tags as well, doubling the size of these events.  And we'd need the old 'r' tags for existing clients until every implementation upgraded, and until all users of old versions of those implementations upgraded, which .... well we would never be able to detect.

So I don't see how it can be fixed at this point. It's a little bit bad currently, but trying to fix this actually makes it worse. 

Unless I'm wrong and I sure hope I'm wrong so please tell me if I'm wrong, I _so_ want to be wrong about this.

**Comment by mikedilger** [comment]
We could instead just put in there that relays don't need to index these particular 'r' tags if they don't want to as we don't expect those indexes to be needed by anyone.

**Comment by arthurfranca** [comment]
> just put in there that relays don't need to index these particular 'r' tags

Good and simple solution, probably the best path to take.

> maybe even upgraded the markers to "inbox" and "outbox"

What if we do it like this:

inbox: `["r", "wss://relay1.example.com", "read", "dm"]`
outbox: `["r", "wss://relay1.example.com", "write", "dm"]`
both: `["r", "wss://relay1.example.com", "", "dm"]`
general: `["r", "wss://relay1.example.com"]`
many special cases: `["r", "wss://relay1.example.com", "", "dm", "other_future_stuff"]`

It prevents breakage of current apps using just "read" and "write".

**Comment by weex** [comment]
> A bit offtopic but.. never understood why 10002 is using `r` tags instead of `relay` tags. Why index all these values bloating relays' DBs?

It might be good to create an issue for this OT thread so folks with the same question can find it easily.

**Comment by weex** [comment]
To the topic, is this same method used elsewhere? Perhaps a reference or two would help the discussion.

**Comment by fiatjaf** [comment]
```
tags: [
  ["inbox", "wss://whatever", "wss://something", ...],
  ["outbox", "wss://blabla", "wss://foobar"]
]
```

**Comment by staab** [comment]
I don't like variadic tags, but @fiatjaf that's a good idea, what about:

```
tags: [
  ["inbox", "wss://whatever"],
  ["inbox", "wss://foobar"],
  ["outbox", "wss://blabla"]
]
```

**Comment by vitorpamplona** [comment]
Since we are here: 

tags: [
  ["public_mentions", "wss://whatever"],  // you can send to it and others can use to download events that cite me
  ["private_inbox", "wss://foobar"],           // you can send to it, but only I can download these events  
  ["outbox", "wss://blabla"].           // Anything I wrote recently is cached here.
  ["archive", "wss://blabla"].           // All my historical events. (find versions of my replaceable events here)
]

**Comment by arthurfranca** [comment]
Yeah clever to skip "r"/"relay" tag name and use the markers directly, considering the event is supposed to have just relay urls.

I like "read"/"write" but I see there's fear people get confused so you guys chose "inbox"/"outbox" as a clearer synonym.

```
tags: [
  ["inbox", "wss://replies-reactions-mentions.dms.xyz"],
  ["inbox", "wss://dms.xyz", "{ \"kinds\": [4, 14] }"], // prefer this one for these kinds
  ["outbox", "wss://feed.xyz"],
  ["outbox", "wss://feed-fallback.xyz"], // second one has lower precedence
]
```

**Comment by fiatjaf** [comment]
Why the hate against variadic args? This is not a function call, it is just a list of items, no need to repeat the key a dozen times, we save space and bandwidth and memory and programs are faster and code is simpler.

Parsing with "variadic args":

```js
let inbox = event.tags.find(t => t[0] === 'inbox')?.slice(1) || []
```

Parsing with a dozen tags:

```js
let inbox = event.tags.filter(t => t[0] === 'inbox').map(t => t[1] || null).filter(r => r) || []
```

**Comment by staab** [comment]
Variadic args make it impossible to add additional information to an entry like in @arthurfranca's example in a backwards-compatible way. Similar problem with positional arguments. Positional arguments couple order with meaning, associativity couples name and meaning.

**Comment by mikedilger** [comment]
I like replacing the "r" with "inbox" and "outbox" tags.

I'm also in favor of ordering such as here: https://github.com/nostr-protocol/nips/pull/822
but only if/when we are making new tags. In this case we can say that the order matters right from the start.

But if we switch to "inbox" and "outbox" tags and maybe some others (we should settle on what those others are), how do clients migrate?  Do we just put "r" tags as well and then in 6 months or so stop putting in the "r" tags?  That would be ok by me as long as we had a deprecation process. We should have a deprecation process where we prominently list things being deprecated on the main page, so developers can keep up with what is important without wading through everything.

**Comment by frnandu** [comment]
Is there anything that prevents tags from being a mixed array of objects and/or arrays, instead of just an array of arrays?
If not, why not something like:
`tags: [
   { "relay": "wss://personal.org", "use_for": ["outbox"], "order": 1 },
   { "relay": "wss://popular.org" : "use_for": ["read","future_tag"] },
   { "relay": "wss://private-inbox.org", "use_for": ["dm"] },
   ["r","wss://old-relay.org","read"] 
]
`
doesn't duplicate relay urls, and can still support old style of arrays with deprecated "r" that clients will still be using for a while...
Existence of optional "order" field would allow for clients that want to support it (for outbox/gossip model of feed fetching) to use it and sort among a specific "use_for".

**Comment by fiatjaf** [comment]
They must be an array of array of strings, for parser sanity in typed languages.

**Comment by frnandu** [comment]
> They must be an array of arrays of strings, for parser sanity in typed languages.

Yeah, I just realized after posting that on nip01 tags are defined as an array of arrays. Sorry for that.
